src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -153,14 +246,34 @@ Node CegisUnif::purifyLemma(Node n,
   /* Recurse */
   unsigned size = n.getNumChildren();
   Kind k = n.getKind();
-  bool fapp = k == APPLY_UF && size > 0 && n[0] == d_candidate;
+  bool fapp = k == APPLY_UF && size > 0
+              && d_candidates.find(n[0]) != d_candidates.end();
   bool childChanged = false;
   std::vector<Node> children;
+  NodeManager* nm = NodeManager::currentNM();
   for (unsigned i = 0; i < size; ++i)
   {
     if (i == 0 && fapp)
     {
-      children.push_back(n[0]);
+      /* Non unif functions are kept as is */
+      if (!d_sygus_unif.usingUnif(n[0]))
+      {
+        children.push_back(n[0]);
+        continue;
+      }
+      std::map<Node, Node>::const_iterator it = d_app_to_purified.find(n);
```

Comment:
I think you need to check cache based on the node *after* it has been purified.  Instead, we are caching based on n which is not necessary in the form f( e, c1...cn ).